File: javascript/refactoring/order.js
class OrderProcessor {
  constructor(user, items, discountCode, shippingAddress, billingAddress) {
    this.user = user;
    this.items = items;
    this.discountCode = discountCode;
    this.shippingAddress = shippingAddress;
    this.billingAddress = billingAddress;
  }

  process() {
    // magic number
    const TAX_RATE = 0.07;
    let total = 0;
    for (let i = 0; i < this.items.length; i++) {
      const item = this.items[i];
      total += item.price * item.quantity;
    }

    // duplicate code with shipping and billing
    if (this.shippingAddress.country === "US") {
      total += 5; // flat shipping US
    } else {
      total += 15; // flat shipping elsewhere
    }

    if (this.billingAddress.country === "US") {
      total += 5;
    } else {
      total += 15;
    }

    if (this.discountCode) {
      if (this.discountCode === "SUMMER10") {
        total = total * 0.90;
      } else if (this.discountCode === "VIP") {
        total = total * 0.80;
      } else {
        total = total * 0.95;
      }
    }

    const tax = total * TAX_RATE;
    total += tax;

    // inline variable example
    if (this.user.age > 18 && this.user.status === "active" && this.user.country === "US") {
      console.log("Ready to process order for active adult US user");
    }

    // long method continues
    this.saveOrder({
      userId: this.user.id,
      items: this.items,
      discount: this.discountCode,
      shipping: this.shippingAddress,
      billing: this.billingAddress,
      total: total
    });

    this.sendConfirmationEmail(this.user.email, total);

    if (this.items.length > 10) {
      this.applyBulkOrderBonus();
    }

    // more logic for inventory update, audit logging, etc.
    for (let i = 0; i < this.items.length; i++) {
      const item = this.items[i];
      // magic number again
      if (item.quantity > 100) {
        console.warn("Large quantity order for item:", item.id);
      }
      this.decreaseInventory(item.id, item.quantity);
    }

    this.logAudit(this.user.id, this.items, total);
  }

  saveOrder(orderData) {
    // imagine DB call
    console.log("Saving order data", orderData);
  }

  sendConfirmationEmail(email, amount) {
    console.log(`Sending confirmation email to ${email} for amount $${amount}`);
  }

  applyBulkOrderBonus() {
    console.log("Applying bulk order bonus");
  }

  decreaseInventory(itemId, qty) {
    console.log(`Decreasing inventory for ${itemId} by ${qty}`);
  }

  logAudit(userId, items, amount) {
    console.log(`Audit log: user ${userId}, items ${items.length}, amount ${amount}`);
  }
}

module.exports = OrderProcessor

File: python/refactoring/order.py
# order_processor.py
class OrderProcessor:
    def __init__(self, user, items, discount_code, shipping_address, billing_address):
        self.user = user
        self.items = items
        self.discount_code = discount_code
        self.shipping_address = shipping_address
        self.billing_address = billing_address

    def process(self):
        TAX_RATE = 0.07
        total = 0
        for item in self.items:
            total += item["price"] * item["quantity"]

        if self.shipping_address["country"] == "US":
            total += 5
        else:
            total += 15

        if self.billing_address["country"] == "US":
            total += 5
        else:
            total += 15

        if self.discount_code:
            if self.discount_code == "SUMMER10":
                total = total * 0.90
            elif self.discount_code == "VIP":
                total = total * 0.80
            else:
                total = total * 0.95

        tax = total * TAX_RATE
        total += tax

        if self.user["age"] > 18 and self.user["status"] == "active" and self.user["country"] == "US":
            print("Ready to process order for active adult US user")

        self.save_order({
            "userId": self.user["id"],
            "items": self.items,
            "discount": self.discount_code,
            "shipping": self.shipping_address,
            "billing": self.billing_address,
            "total": total
        })

        self.send_confirmation_email(self.user["email"], total)

        if len(self.items) > 10:
            self.apply_bulk_order_bonus()

        for item in self.items:
            if item["quantity"] > 100:
                print("WARNING: Large quantity order for item:", item["id"])
            self.decrease_inventory(item["id"], item["quantity"])

        self.log_audit(self.user["id"], self.items, total)

    def save_order(self, order_data):
        print("Saving order data", order_data)

    def send_confirmation_email(self, email, amount):
        print(f"Sending confirmation email to {email} for amount ${amount}")

    def apply_bulk_order_bonus(self):
        print("Applying bulk order bonus")

    def decrease_inventory(self, item_id, qty):
        print(f"Decreasing inventory for {item_id} by {qty}")

    def log_audit(self, user_id, items, amount):
        print(f"Audit log: user {user_id}, items {len(items)}, amount {amount}")
